### ISSUE TITLE
PDF support does not work on Pi devices

### ISSUE THREAD

**Issue by majora2007** [Kareadita/Kavita#468]
It seems docnet does not support pdfium's ARM build, thus Pi users who use the ARM build cannot read PDF files. 

It seems we need to fork pdfium, build an arm version and add it to docnet and use the custom version. 

https://pdfium.googlesource.com/pdfium/

**Comment by majora2007** [comment]
NetVips can handle PDFs through Poppler. Need to validate how to include the proper binaries with our build.
https://github.com/kleisauke/net-vips/issues/11

```
public void ExtractPdfImages(string fileFilePath, string targetDirectory)
        {
            DirectoryService.ExistOrCreate(targetDirectory);

            //using var docReader = DocLib.Instance.GetDocReader(fileFilePath, new PageDimensions(1080, 1920));
            //var pages = docReader.GetPageCount();
            var pages = GetPdfPageCount(fileFilePath);

            using var stream = StreamManager.GetStream("BookService.GetPdfPage");
            for (var pageNumber = 0; pageNumber < pages; pageNumber++)
            {
                GetPdfPage(fileFilePath, pageNumber, stream);
                //File.WriteAllBytes(Path.Combine(targetDirectory, "Page-" + pageNumber + ".png"), stream.ToArray());
                using var fileStream = File.Create(Path.Combine(targetDirectory, "Page-" + pageNumber + ".png"));
                stream.Seek(0, SeekOrigin.Begin);
                stream.CopyTo(fileStream);
            }
        }

private static void GetPdfPage(string filename, int pageNumber, Stream stream)
        {
            using var image = NetVips.Image.Pdfload(filename, pageNumber );
            image.PngsaveStream(stream);
        }
rivate static int GetPdfPageCount(string filePath)
        {
            try
            {
                using var fs = new FileStream(filePath, FileMode.Open, FileAccess.Read);
                using var bs = new BufferedStream(fs);
                using var reader = new StreamReader(bs);

                var pdfText = reader.ReadToEnd();
                var matches = Parser.Parser.PdfPageMarker.Matches(pdfText);

                return matches.Count;
            }
            catch (Exception)
            {
                /* Swallow exception */
            }

            return 0;
        }

```

**Comment by thib1984** [comment]
I'm not sure but, it can explain why the docker image does not work in mi pi device (but is ok in WSL?) for the pdf support?

**Comment by majora2007** [comment]
@thib1984 The issue that I'm seeing is the underlying library we choose (docnet which wraps pdfium which is the chrome pdf renderer) does not have arm based binaries of pdfium, which is why it fails. I would assume WSL still uses the windows pdfium build and the translation layer makes it seemless. 

This issue is a high priority item and if you see the Projects page, we have assigned it a release on it's own very close to the current version.

**Comment by ChristofferGreen** [comment]
Has this issue been descoped from the release roadmap?

**Comment by majora2007** [comment]
Yes. It's a serious pain from the research I've done that revolves around
building underlying libraries. For the time being, I'm leaving it to
someone else, plus I don't have a pi, so even more difficult for me to
tackle.

If you're interested, let me know. I can give you the info and work on
discord with you.

**Comment by ChristofferGreen** [comment]
Ah. Sure, I can have a look since the mangapy patch is done now. I would love to have pdf support on the pi (thats where I run Kavita) so would be interesting for me to look into, though not sure how hard it will be.

**Comment by ChristofferGreen** [comment]
![image](https://user-images.githubusercontent.com/248162/132653424-f18a2593-e489-421d-99f7-634643cd1fa4.png)
First pdf rendered on my raspberry pi. Still a bit of work todo on figuring out how to get a bunch of github repos to cooperate so kavita can grab a pdfium.so binary.

**Comment by majora2007** [comment]
A PR has been opened on a downstream library. This is the first step to support.
https://github.com/bblanchon/pdfium-binaries/pull/47

**Comment by ChristofferGreen** [comment]
Status update: The patch has been merged in to the pdfium-binaries repo and we are now waiting for the automatic builds to produce binaries, this may take a few days or a week or two.

After that, we need to update the docnet repo with the pdfium-binaries library.

**Comment by ChristofferGreen** [comment]
Status update:
There are now builds of the library for 32-bit arm: https://github.com/bblanchon/pdfium-binaries/releases/latest/download/pdfium-linux-arm.tgz
If anyone wants to they can simply replace the pdfium.so file in Kavita with that one and pdf support should work.

I have also submitted a pull-request to the docket library to pull in this file. Docnet is the library Kavita uses to read pdfs. https://github.com/GowenGit/docnet/pull/50

**Comment by majora2007** [comment]
Once the underlying PR is merged, we need to put information on the wiki to outline that users not using Docker (aka PI users), will need to install libgdiplus on their systems for PDF rendering to work. 

It does not come natively on linux, but we use it for rendering PDF images out.

**Comment by majora2007** [comment]
This will be moved to the next release, however I will manually update builds for the v0.4.6 release to have the updated library.
